src/backend/distributed/executor/repartition_logic.c

```diff
@@ -0,0 +1,252 @@
+ * - It generates all the tasks by descending in the tasks tree. Note that each task
+ *  has a dependedTaskList.
+ * - It generates FetchTask queryStrings with the MapTask queries. It uses the first replicate to
+ *  fetch data when replication factor is > 1. Note that if a task fails in any replica adaptive executor
+ *  gives an error, so if we come to a fetchTask we know for sure that its dependedMapTask is executed in all
+ *  replicas.
+ * - It creates schemas in each worker in a single transaction to store intermediate results.
+ * - It iterates all tasks and finds the ones whose dependencies are already executed, and executes them with
+ *  adaptive executor logic.
+ *
+ *
+ * Repartition queries do not begin a transaction even if we are in
+ * a transaction block. As we dont begin a transaction, they wont see the
+ * DDLs that happened earlier in the transaction because we dont have that
+ * transaction id with repartition queries. Therefore we error in this case.
+ *
+ * Copyright (c) Citus Data, Inc.
+ */
+
+#include "postgres.h"
+#include "access/hash.h"
+#include "distributed/hash_helpers.h"
+
+#include "distributed/directed_acylic_graph_execution_logic.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/adaptive_executor.h"
+#include "distributed/worker_manager.h"
+#include "distributed/multi_server_executor.h"
+#include "distributed/repartition_logic.h"
+#include "distributed/worker_transaction.h"
+#include "distributed/worker_manager.h"
+#include "distributed/transaction_management.h"
+#include "distributed/multi_task_tracker_executor.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/transmit.h"
+
+
+static void FillTaskGroups(List **allTasks, List **mergeTasks);
+static List * CreateTemporarySchemas(List *mergeTasks);
+static List * CreateJobIds(List *mergeTasks);
+static void CreateSchemasOnAllWorkers(char *createSchemasCommand);
+static char * GenerateCreateSchemasCommand(List *jobIds);
+static char * GenerateCommand(List *jobIds, char *templateCommand);
+static bool DoesJobIDExist(List *jobIds, uint64 jobId);
+static void SendCommandToAllWorkers(List *commandList);
+static char * GenerateDeleteJobsCommand(List *jobIds);
+static void RemoveTempJobDirs(List *jobIds);
+
+
+/*
+ * ExecuteDependedTasks executes all tasks except the top level tasks
+ * in order from the task tree. At a time, it can execute different tasks from
+ * different jobs.
+ */
+void
+ExecuteDependedTasks(List *topLevelTasks)
+{
+	List *mergeTasks = NIL;
+
+
+	EnsureNoModificationsHaveBeenDone();
+
+	List *allTasks = TaskAndExecutionList(topLevelTasks);
+
+	FillTaskGroups(&allTasks, &mergeTasks);
+
+	List *jobIds = CreateTemporarySchemas(mergeTasks);
+
+	ExecuteTasksInDependencyOrder(allTasks, topLevelTasks);
+
+	RemoveTempJobDirs(jobIds);
+}
+
+
+/*
+ * FillTaskGroups iterates all tasks and creates a group for mergeTasks.
+ */
+static void
+FillTaskGroups(List **allTasks, List **mergeTasks)
```

Comment:
This API seems very weird. What about `List * ExtractMergeTasks(List *allTasks)`? If we need to concatenate the result then use list_concat.

Reply:
My bad, it is a leftover name from a previous version. Now It makes no sense. I will change it to ExtractMergeTasks.